### DISCUSSION TITLE
Loos 4.0

### DISCUSSION THREAD

**Discussion by tromo** [GrossfieldLab/loos#86] [General]
We are pleased to announce the release of LOOS-4.0.  If you have been working from the main branch, there should be minimal change in how you use LOOS.  However, behind the scenes, there have been many changes.  In particular, we are now using `cmake` rather than `scons` for building LOOS.  In addition, PyLOOS is installed via `pip` into your Python (or conda) environment.  There are also several new tools.

Within the next week or two, we expect to release a version installable via conda-forge, which will make installation much simpler.

<hr /><em>This discussion was created from the release <a href='https://github.com/GrossfieldLab/loos/releases/tag/v4.0.0'>Loos 4.0</a>.</em>

**Comment by agrossfield** [comment]
Like the above note says, the biggest change in LOOS 4.0 is the infrastructure behind building and installing. For those ready to try LOOS 4.0, we have different suggestions for how to get it, depending on your use case:

- If you already have a working git clone of the main branch, especially if it's in a conda environment, I'd suggest recloning a new directory and following the build instructions in INSTALL.md. The only new requirement is cmake, which you may need to add to your conda environment. Alternatively, you can run `scons -c; scons -c config` to clean out the current directory and do a `git pull` to bring down the latest version.
- If your main plan is to use existing tools (c++ or python), I'd suggest waiting for the conda-forge recipe to get accepted (we'll announce it here when it happens, hopefully in the next week or two). This will let you avoid a big download of conda tools and skip building loos for yourself.
- The same is true if you anticipate writing new python tools. 
- If you anticipate either writing new c++ tools or modifying the loos library itself, you'll want to have a full clone of loos. Your best bet is to clone from GitHub and (assuming you've already got conda set up) build using the included conda_build.sh script. The script will create a new environment, install the dependencies, then build and install loos.
- We have tested building using native system libraries instead of conda on several common linux distributions. Instructions are in INSTALL.md, but YMMV if you use anything we haven't tested.

**Comment by agrossfield** [comment]
Note for Mac users with ARM chips: right now, the conda build of loos 4.0 is broken. The issue is broken cmake packaging for libnetcdf -- we've notified the package maintainers, but until they fix it none of the conda install paths will work correctly; this is also the reason why the conda_build pins the version of libnetcdf, because it's broken on all OSX platforms. The difference is there is no working version for arm. 

We apologize for the inconvenience. We are monitoring the situation and will post updates here. 

If you are impatient to try loos 4.0 and don't want to wait for the libnetcdf packagers to fix their problem, you can probably fix it yourself in a half-assed sort of way:

- Look at conda_build.ah, and run the commands that set up the conda env and install the necessary packages, including libnetcf.
- Go to $CONDA_PREFIX/lib/cmake/netCDF and open netCDFTargets.cmake in your favorite text editor.
- Look for the line that starts with `INTERFACE_LINK_LIBRARIES` followed by a string with a bunch of paths. Buried in the middle of it is a bit that looks like `/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk/usr/lib/libm.tbd`, where the packager accidentally referenced the Xcode libm instead of the one within the conda env. I suspect that removing that path will fix the problem. If it doesn't and for some reason libnetcdf actually needs libm, you might need to install the openlibm package so there's a libm inside conda.
- Once you've made this edit, follow the install instructions in INSTALL.md inside loos for building under conda:
```
mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX ..
cmake --build .
cmake --install .
``` 

I'm hoping the netcdf packagers will fix their problems soon so that you don't need to do this.

**Reply by agrossfield** [reply]
The libnetcdf 4.9.0 package now works correctly, and the fix is back-ported to 4.8.1 (so those packages should work in a few hours).
